src/runtime_src/core/edge/drm/zocl/edge/zocl_hwctx.c

```diff
@@ -0,0 +1,460 @@
+        /* Only start kernel command supports timestamps */
+        struct ert_start_kernel_cmd *scmd;
+        struct cu_cmd_state_timestamps *ts;
+
+        scmd = (struct ert_start_kernel_cmd *)ecmd;
+        ts = ert_start_kernel_timestamps(scmd);
+        ts->skc_timestamps[ERT_CMD_STATE_NEW] = xcmd->timestamp[KDS_NEW];
+        ts->skc_timestamps[ERT_CMD_STATE_QUEUED] = xcmd->timestamp[KDS_QUEUED];
+        ts->skc_timestamps[ERT_CMD_STATE_RUNNING] = xcmd->timestamp[KDS_RUNNING];
+        ts->skc_timestamps[ecmd->state] = xcmd->timestamp[status];
+    }
+
+    ZOCL_DRM_GEM_OBJECT_PUT_UNLOCKED(xcmd->gem_obj);
+
+    if (xcmd->cu_idx >= 0)
+        client_stat_inc(client, xcmd->hw_ctx_id, c_cnt[xcmd->cu_idx]);
+
+    atomic_inc(&client->event);
+    wake_up_interruptible(&client->waitq);
+}
+
+static struct kds_client_cu_ctx *
+zocl_get_hw_cu_ctx(struct kds_client_hw_ctx *kds_hw_ctx, int cu_idx)
+{
+    struct kds_client_cu_ctx *kds_cu_ctx = NULL;
+    bool found = false;
+
+    list_for_each_entry(kds_cu_ctx, &kds_hw_ctx->cu_ctx_list, link) {
+        if (kds_cu_ctx->cu_idx == cu_idx) {
+            found = true;
+            break;
+        }
+    }
+
+    if (found)
+        return kds_cu_ctx;
+    return NULL;
+}
+
+static int
+check_for_open_hw_cu_ctx(struct drm_zocl_dev *zdev, struct kds_client *client, struct kds_command *xcmd)
+{
+    struct kds_client_hw_ctx *kds_hw_ctx = NULL;
+    int first_cu_idx = -EINVAL;
+    u32 mask = 0;
+    int i, j;
+    int ret = 0;
+
+    /* i for iterate masks, j for iterate bits */
+    for (i = 0; i < xcmd->num_mask; ++i) {
+        if (xcmd->cu_mask[i] == 0)
+            continue;
+
+        mask = xcmd->cu_mask[i];
+        for (j = 0; mask > 0; ++j) {
+            if (!(mask & 0x1)) {
+                mask >>= 1;
+                continue;
+            }
+
+            first_cu_idx = i * sizeof(u32) + j;
+            goto out;
+        }
+    }
+
+out:
+    if (first_cu_idx < 0)
+        return -EINVAL;
+
+    mutex_lock(&client->lock);
+    kds_hw_ctx = kds_get_hw_ctx_by_id(client, xcmd->hw_ctx_id);
+    if (!kds_hw_ctx) {
+        mutex_unlock(&client->lock);
+        return -EINVAL;
+    }
+
+    if (zocl_get_hw_cu_ctx(kds_hw_ctx, first_cu_idx) != NULL)
+        ret = 0;
+    else
+        ret = -EINVAL;
+
+    mutex_unlock(&client->lock);
+    return ret;
+}
+
+int zocl_hw_ctx_execbuf(struct drm_zocl_dev *zdev, struct drm_zocl_hw_ctx_execbuf *drm_hw_ctx_execbuf, struct drm_file *filp)
+{
+    struct drm_gem_object *gem_obj = NULL;
+    struct drm_device *dev = zdev->ddev;
+    struct kds_client *client = filp->driver_priv;
+    struct drm_zocl_bo *zocl_bo = NULL;
+    struct ert_packet *ecmd = NULL;
+    struct kds_command *xcmd = NULL;
+    int ret = 0;
+
+    if (zdev->kds.bad_state) {
+        DRM_ERROR("%s: KDS is in bad state", __func__);
+        return -EDEADLK;
+    }
+    gem_obj = zocl_gem_object_lookup(dev, filp, drm_hw_ctx_execbuf->exec_bo_handle);
```

Comment:
When you use drm_gem_object_lookup, you typically need to release the reference to the GEM object once you’re done with it. This is where drm_gem_object_put_unlocked comes into play.

So , when ever you are erroring out you should use ZOCL_DRM_GEM_OBJECT_PUT_UNLOCKED.
Check this function.

Reply:
Hi Saif, releasing the ref to the GEM object properly before erroring out. Made the necessary changes on latest push. Thanks